packages/core/js-client/src/particle/Particle.ts

```diff
@@ -14,145 +14,179 @@
  * limitations under the License.
  */
 
-import { atob, fromUint8Array, toUint8Array } from 'js-base64';
-import { CallResultsArray } from '@fluencelabs/avm';
-import { v4 as uuidv4 } from 'uuid';
-import { Buffer } from 'buffer';
-import { IParticle } from './interfaces.js';
-import { concat } from 'uint8arrays/concat';
-import { numberToLittleEndianBytes } from '../util/bytes.js';
-import { KeyPair } from '../keypair/index.js';
-import { unmarshalPublicKey } from '@libp2p/crypto/keys';
+import { Buffer } from "buffer";
+
+import { CallResultsArray } from "@fluencelabs/avm";
+import { fromUint8Array, toUint8Array } from "js-base64";
+import { concat } from "uint8arrays/concat";
+import { v4 as uuidv4 } from "uuid";
+import { z } from "zod";
+
+import { KeyPair } from "../keypair/index.js";
+import { numberToLittleEndianBytes } from "../util/bytes.js";
+
+import { IParticle } from "./interfaces.js";
+
+const particleSchema = z.object({
+  id: z.string(),
+  timestamp: z.number().positive(),
+  script: z.string(),
+  data: z.string(),
+  ttl: z.number().positive(),
+  init_peer_id: z.string(),
+  signature: z.array(z.number()),
+});
 
 export class Particle implements IParticle {
-    constructor(
-        public readonly id: string,
-        public readonly timestamp: number,
-        public readonly script: string,
-        public readonly data: Uint8Array,
-        public readonly ttl: number,
-        public readonly initPeerId: string,
-        public readonly signature: Uint8Array
-    ) {}
-
-    static async createNew(script: string, initPeerId: string, ttl: number, keyPair: KeyPair): Promise<Particle> {
-        const id = uuidv4();
-        const timestamp = Date.now();
-        const message = buildParticleMessage({ id, timestamp, ttl, script });
-        const signature = await keyPair.signBytes(message);
-        return new Particle(id, Date.now(), script, Buffer.from([]), ttl, initPeerId, signature);
+  constructor(
+    readonly id: string,
+    readonly timestamp: number,
+    readonly script: string,
+    readonly data: Uint8Array,
+    readonly ttl: number,
+    readonly initPeerId: string,
+    readonly signature: Uint8Array,
+  ) {}
+
+  static async createNew(
+    script: string,
+    initPeerId: string,
+    ttl: number,
+    keyPair: KeyPair,
+  ): Promise<Particle> {
+    const id = uuidv4();
+    const timestamp = Date.now();
+    const message = buildParticleMessage({ id, timestamp, ttl, script });
+    const signature = await keyPair.signBytes(message);
+    return new Particle(
+      id,
+      Date.now(),
+      script,
+      Buffer.from([]),
+      ttl,
+      initPeerId,
+      signature,
+    );
+  }
+
+  static fromString(str: string): Particle {
+    const json = JSON.parse(str);
+
+    const res = particleSchema.safeParse(json);
+
+    if (!res.success) {
+      throw new Error(`Particle format invalid. Given: ${str}`);
     }
```

Comment:
Are erros that zod gives not good? Why don't you display them? If some problem accurs for some reason at least this way we would know what format is actually valid, what is expected. Here we can only see what we got, but not what we exepect to get

Reply:
Right. Fixed